grpc/alteration_percent_test.go

```diff
@@ -0,0 +1,587 @@
+				},
+				{
+					ErrorToReturn:    "ALREADY_EXISTS",
+					OverrideToReturn: "",
+					QueryPercent:     int32(30),
+				},
+				{
+					ErrorToReturn:    "",
+					OverrideToReturn: "{}",
+					QueryPercent:     int32(40),
+				},
+			}
+
+			var err error
+			mapping, err = GetAlterationToPercentAffected(alterationSpecs, "endpointname")
+			Expect(err).To(BeNil())
+		})
+
+		It("should create a mapping with 3 elements", func() {
+			Expect(len(mapping)).To(Equal(3))
+		})
+
+		It("should create a mapping with correct configs", func() {
+			var altCfg AlterationConfiguration
+
+			altCfg = AlterationConfiguration{
+				ErrorToReturn:    "CANCELED",
+				OverrideToReturn: "",
+			}
+			pct_cancelled, ok_cancelled := mapping[altCfg]
+			Expect(ok_cancelled).To(BeTrue())
+			Expect(pct_cancelled).To(Equal(PercentSlot(20)))
+
+			altCfg = AlterationConfiguration{
+				ErrorToReturn:    "ALREADY_EXISTS",
+				OverrideToReturn: "",
+			}
+			pct_exists, ok_exists := mapping[altCfg]
+			Expect(ok_exists).To(BeTrue())
+			Expect(pct_exists).To(Equal(PercentSlot(30)))
+
+			altCfg = AlterationConfiguration{
+				ErrorToReturn:    "",
+				OverrideToReturn: "{}",
+			}
+			pct_emptyret, ok_emptyret := mapping[altCfg]
+			Expect(ok_emptyret).To(BeTrue())
+			Expect(pct_emptyret).To(Equal(PercentSlot(40)))
+		})
+	})
+
+	Context("with one alterations with too many fields specified", func() {
+		var err error
+
+		BeforeEach(func() {
+			alterationSpecs = []*pb.AlterationSpec{
+				{
+					ErrorToReturn:    "CANCELED",
+					OverrideToReturn: "{}",
+				},
+			}
+
+			_, err = GetAlterationToPercentAffected(alterationSpecs, "endpointname")
+		})
+
+		It("should fail", func() {
+			Expect(err.Error()).To(Equal("Cannot execute SendDisruption without specifying either ErrorToReturn or OverrideToReturn for all target endpoints"))
+		})
+	})
+
+	Context("with one alterations with too few fields specified", func() {
+		var err error
+
+		BeforeEach(func() {
+			alterationSpecs = []*pb.AlterationSpec{
+				{
+					ErrorToReturn:    "",
+					OverrideToReturn: "",
+				},
+			}
+
+			_, err = GetAlterationToPercentAffected(alterationSpecs, "endpointname")
+		})
+
+		It("should fail", func() {
+			Expect(err.Error()).To(Equal("Cannot execute SendDisruption where ErrorToReturn or OverrideToReturn are both specified for a target endpoints"))
+		})
+	})
+
+	Context("with three alterations which are more than 100", func() {
+		BeforeEach(func() {
+			alterationSpecs = []*pb.AlterationSpec{
+				{
+					ErrorToReturn:    "CANCELED",
+					OverrideToReturn: "",
+					QueryPercent:     int32(50),
+				},
+				{
+					ErrorToReturn:    "ALREADY_EXISTS",
+					OverrideToReturn: "",
+					QueryPercent:     int32(50),
+				},
+				{
+					ErrorToReturn:    "",
+					OverrideToReturn: "{}",
+					QueryPercent:     int32(50),
+				},
+			}
+
+			var err error
+			mapping, err = GetAlterationToPercentAffected(alterationSpecs, "endpointname")
+			Expect(err).To(BeNil())
```

Comment:
It would only return an error if the alterationSpecs are invalid. In this case, it does not return an error because `endpointname` is only used to print errors. This "smells" bad to pass a parameter just for the error state, but I don't really know how to error without the name of the endpoint, how can the user correct their mistake? Should I be handling the error outside of the function?

Reply:
I think I understand? But if this is a situation we expect to never be in due to validation, why test that it works?

Reply:
Tell me if this makes sense:

The (in)correctness of having over 100% is checked in `percent_to_alteration` file, this is just explicitly showing that it's valid but we will never access more than 100%. Few rows under:

> // intuition here as that the service will never pick a number larger than 100 randomly
> // so this return value never gets triggered, but the function does not error out